File: 258.add-digits.cpp
/*
 * @lc app=leetcode id=258 lang=cpp
 *
 * [258] Add Digits
 */

#include <bits/stdc++.h>

#include <cassert>

using namespace std;

// @lc code=start
class Solution {
 public:
  int addDigits(int num) {
    if (!num) return num;
    return (num % 9) ? (num % 9) : 9;
  }
  int addDigitsNormal(int num) {
    while (num >= 10) {
      int val = num;
      int sum = 0;
      while (val) {
        sum += val % 10;
        val /= 10;
      }
      num = sum;
    }
    return num;
  }
};
// @lc code=end

int main() {
  Solution s;
  assert(s.addDigits(10) == 1);
  assert(s.addDigits(38) == 2);
  assert(s.addDigits(0) == 0);
  return 0;
}

File: 258.add-digits.rs
/*
 * @lc app=leetcode id=258 lang=rust
 *
 * [258] Add Digits
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn add_digits(num: i32) -> i32 {
        let mut num = num;
        while num >= 10 {
            let mut work = num;
            num = 0;
            while work > 0 {
                num += work % 10;
                work /= 10;
            }
        }
        num
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::add_digits(38), 2);
}

#[test]
fn test2() {
    assert_eq!(Solution::add_digits(0), 0);
}
